cmd/oras/discover.go

```diff
@@ -0,0 +1,207 @@
+	root := tree.New(opts.targetRef)
+	desc, err := repo.Resolve(ctx, repo.Reference.ReferenceOrDefault())
+	if err != nil {
+		return err
+	}
+	var refs []artifactspec.Descriptor
+	if opts.outputType == "tree" {
+		refs, err = fetchAllReferrers(ctx, repo, desc, opts.artifactType, root)
+	} else {
+		refs, err = fetchReferrers(ctx, repo, desc, opts.artifactType)
+	}
+	if err != nil {
+		return err
+	}
+
+	switch opts.outputType {
+	case "tree":
+		tree.Print(root)
+	case "json":
+		printDiscoveredReferencesJSON(desc, refs)
+	default:
+		fmt.Println("Discovered", len(refs), "artifacts referencing", opts.targetRef)
+		fmt.Println("Digest:", desc.Digest)
+		if len(refs) > 0 {
+			fmt.Println()
+			printDiscoveredReferencesTable(refs, opts.Verbose)
+		}
+	}
+	return nil
+}
+
+func fetchReferrers(ctx context.Context, repo *remote.Repository, desc ocispec.Descriptor, artifactType string) ([]artifactspec.Descriptor, error) {
+	var results []artifactspec.Descriptor
+	err := repo.Referrers(ctx, desc, artifactType, func(referrers []artifactspec.Descriptor) error {
+		results = append(results, referrers...)
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	return results, nil
+}
+
+func fetchAllReferrers(ctx context.Context, repo *remote.Repository, desc ocispec.Descriptor, artifactType string, node *tree.Node) ([]artifactspec.Descriptor, error) {
+	var results []artifactspec.Descriptor
+	results, err := fetchReferrers(ctx, repo, desc, artifactType)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, r := range results {
+		// Find all indirect referrers
+		referrerNode := node.AddPath(r.ArtifactType, r.Digest)
+		nestedReferrers, err := fetchAllReferrers(
+			ctx, repo,
+			ocispec.Descriptor{
+				Digest:    r.Digest,
+				Size:      r.Size,
+				MediaType: r.MediaType,
+			},
+			artifactType, referrerNode)
+		if err != nil {
+			return nil, err
+		}
+		results = append(results, nestedReferrers...)
+	}
+	return results, nil
+}
+
+func printDiscoveredReferencesTable(refs []artifactspec.Descriptor, verbose bool) {
+	typeNameTitle := "Artifact Type"
+	typeNameLength := len(typeNameTitle)
+	for _, ref := range refs {
+		if length := len(ref.ArtifactType); length > typeNameLength {
+			typeNameLength = length
+		}
+	}
+
+	print := func(key string, value interface{}) {
+		fmt.Println(key, strings.Repeat(" ", typeNameLength-len(key)+1), value)
+	}
+
+	print(typeNameTitle, "Digest")
+	for _, ref := range refs {
+		print(ref.ArtifactType, ref.Digest)
+		if verbose {
+			printJSON(ref)
+		}
+	}
+}
+
+func printDiscoveredReferencesJSON(desc ocispec.Descriptor, refs []artifactspec.Descriptor) {
+	type referrerDesc struct {
+		Digest    digest.Digest `json:"digest"`
+		MediaType string        `json:"mediaType"`
+		Artifact  string        `json:"artifactType"`
+		Size      int64         `json:"size"`
+	}
+	output := struct {
+		// https://github.com/oras-project/artifacts-spec/blob/main/manifest-referrers-api.md#artifact-referrers-api-results
```

Comment:
```suggestion
		// Reference: https://github.com/oras-project/artifacts-spec/blob/v1.0.0-rc.1/manifest-referrers-api.md#artifact-referrers-api-results
```

Reply:
We might want to write documentation for `printDiscoveredReferencesJSON` and move this reference there.